src/Nethermind/Nethermind.AuRa/Validators/ContractValidator.cs

```diff
@@ -0,0 +1,293 @@
+            {
+                _pendingValidators = new PendingValidators(block.Number, potentialValidators);
+                _validatedSinceFinalized.Add(new ValidationStamp(block));
+            }
+        }
+
+        private Address[] LoadValidatorsFromContract(Block block)
+        {
+            ValidatorContract.InvokeTransaction(block.Header, _transactionProcessor, ValidatorContract.GetValidators(), Output);
+
+            var validators = ValidatorContract.DecodeAddresses(Output.ReturnValue);
+            if (validators.Length == 0)
+            {
+                throw new AuRaException("Failed to initialize validators list.");
+            }
+
+            _validators = validators.ToHashSet();
+            return validators;
+        }
+
+
+        private void CreateSystemAccount()
+        {
+            if (!_stateProvider.AccountExists(Address.SystemUser))
+            {
+                _stateProvider.CreateAccount(Address.SystemUser, UInt256.Zero);
+                _stateProvider.Commit(Homestead.Instance);
+            }
+        }
+        
+        private void ClearPendingValidators()
+        {
+            _pendingValidators = null;
+            _validatedSinceFinalized.Clear();
+        }
+
+        private class PendingValidators
+        {
+            public PendingValidators(long blockNumber, Address[] addresses)
+            {
+                BlockNumber = blockNumber;
+                Addresses = addresses;
+            }
+
+            public Address[] Addresses { get; }
+            
+            public long BlockNumber { get; }
+        }
+        
+        private class ValidationStamp : IEquatable<ValidationStamp>
+        {
+
+            public ValidationStamp(long blockNumber, Address address)
+            {
+                BlockNumber = blockNumber;
+                Address = address;
+            }
+
+            public long BlockNumber { get; }
+            
+            public Address Address { get; }
+
+            public ValidationStamp(Block block) : this(block.Number, block.Beneficiary) { }
+
+            public bool Equals(ValidationStamp other)
+            {
+                if (ReferenceEquals(null, other)) return false;
+                if (ReferenceEquals(this, other)) return true;
+                return Equals(Address, other.Address);
+            }
+            
+            public override bool Equals(object obj)
+            {
+                if (ReferenceEquals(null, obj)) return false;
+                if (ReferenceEquals(this, obj)) return true;
+                if (obj.GetType() != GetType()) return false;
+                return Equals((ValidationStamp) obj);
+            }
+
+            public override int GetHashCode()
+            {
+                return Address?.GetHashCode() ?? 0;
+            }
+        }
+        
+        private class ValidationStampCollection
+        {
+            private readonly ISet<ValidationStamp> _set;
```

Comment:
https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.sortedset-1?view=netframework-4.8

Reply:
Yes, that was my initial idea but potentially lower performance O(logn) vs O(1) with current implementation. A little better memory footprint though.